[locid_transform] Optimize canonicalize operation

zbraniecki opened the pull request:
This is an initial dust off of the `canonicalize` operation with a focus on performance optimizations.

It is intended to partially target #805, but it is more generic and may leave some specific tailoring for later.

zbraniecki commented:
I think this is ready for initial review.

```
uncanonicalized/clone   time:   [1.6252 µs 1.6284 µs 1.6316 µs]
                        change: [-0.6316% -0.2661% +0.1023%] (p = 0.16 > 0.05)
                        No change in performance detected.
uncanonicalized/canonicalize
                        time:   [11.184 µs 11.205 µs 11.231 µs]
                        change: [-40.763% -40.594% -40.444%] (p = 0.00 < 0.05)
                        Performance has improved.

canonicalized/clone     time:   [627.23 ns 628.84 ns 630.63 ns]
                        change: [-0.2113% +0.1189% +0.4632%] (p = 0.49 > 0.05)
                        No change in performance detected.
canonicalized/canonicalize
                        time:   [1.2967 µs 1.3001 µs 1.3038 µs]
                        change: [-8.9216% -8.6201% -8.3321%] (p = 0.00 < 0.05)
                        Performance has improved.

likelysubtags/maximize  time:   [1.2064 µs 1.2083 µs 1.2102 µs]
                        change: [+0.0356% +0.3573% +0.6811%] (p = 0.03 < 0.05)
                        Change within noise threshold.
```

zbraniecki commented:
@sffc how should I handle change to the data?

zbraniecki commented:
@dminor - would you have time to review the logic changes in this PR? I will work with @sffc on the data updates, but I think you're the best person to check my updates to alias canonicalization logic.

sffc commented on components/locid_transform/src/provider.rs:
```diff
@@ -61,6 +62,26 @@ pub struct StrStrPair<'a>(
+)]
+/// A triplet of strings with a EncodeAsVarULE implementation.
+pub struct StrStrStrPair<'a>(
+    pub Language,
```
Because `StrStrStrPairVarULE` doesn't have the same binary representation as `StrStrPairVarULE`, this isn't going to be compatible across Postcard binaries. It will need a new `AliasesV2`.

dminor commented:
> @dminor - would you have time to review the logic changes in this PR? I will work with @sffc on the data updates, but I think you're the best person to check my updates to alias canonicalization logic.

It's been so long since I've looked at this code that I no longer feel that comfortable reviewing the changes. I think it would be best to have someone who is more actively involved in day to day ICU4X development have a look, it would be good to have someone else on the team become familiar with locale canonicalization.

zbraniecki commented:
rebase on top of main. There's one concerning benchmark result now where we regress maximize by 4%, I need to debug it before I move to AliasesV2 and request full review.

zbraniecki commented:
I'm also failing debug assert based on baked data:
```
thread 'test_canonicalize' panicked at 'assertion failed: start <= end', /Users/zibi/projects/icu4x/utils/zerovec/src/varzerovec/components.rs:295
```

I did use `./tools/testdata-scripts` to regenerate testdata.

Can I get a pointer at how to debug it? @sffc @robertbastian ?

STR: 
1) Clone my branch
2) go to `./components/locid_transform`
3) `cargo test --all-features`

zbraniecki commented:
I rebased onto old `main`. Updated to current and there's no more `./tools/testdata-scripts`.

I'm still not sure how to find out why the varzerovec fails.

zbraniecki commented:
New benchmarks:

```
uncanonicalized/clone   time:   [2.2011 µs 2.2052 µs 2.2099 µs]
                        change: [-0.2354% +0.0482% +0.3407%] (p = 0.75 > 0.05)
                        No change in performance detected.
uncanonicalized/canonicalize
                        time:   [11.489 µs 11.503 µs 11.518 µs]
                        change: [-43.123% -42.987% -42.855%] (p = 0.00 < 0.05)
                        Performance has improved.

canonicalized/clone     time:   [848.29 ns 850.62 ns 853.03 ns]
                        change: [-0.2430% +0.0619% +0.3581%] (p = 0.69 > 0.05)
                        No change in performance detected.
canonicalized/canonicalize
                        time:   [1.6420 µs 1.6578 µs 1.6724 µs]
                        change: [-7.2969% -6.7797% -6.2620%] (p = 0.00 < 0.05)
                        Performance has improved.

likelysubtags/maximize  time:   [1.4753 µs 1.4781 µs 1.4812 µs]
                        change: [-2.5625% -2.1804% -1.7588%] (p = 0.00 < 0.05)
                        Performance has improved.
```

Please notice that compared to previous benchmarks all of those are ~30% slower. It seems that something happened in the last months that made the tests much slower, it may be the switch of data provider?

sffc commented:
> Please notice that compared to previous benchmarks all of those are ~30% slower. It seems that something happened in the last months that made the tests much slower, it may be the switch of data provider?

The benchmark dashboard has sadly been offline for the last ~2 months so I can't say for sure; might be worth bisecting. Are you sure the benchmarks are taken on the same device both times?

sffc commented:
I bisected it. The culprit commit seems to be: 2c111d43c1ed5e695de60936f182b41ef248e36b

Apparently Rust 1.71 is 30% slower than Rust 1.70

zbraniecki commented:
@sffc do I need to continue packaging AliasesV1 or just being able to consume is sufficient?

sffc commented:
> @sffc do I need to continue packaging AliasesV1 or just being able to consume is sufficient?

The design is that you introduce a _new_ type AliasesV2. It does not replace AliasesV1, which should still work in all contexts _except_ the `_unstable` constructor (which is unstable).

sffc commented:
It's a good exercise on how to update our data structs in a forwards-compatible way. It shouldn't take _too much_ time; basically just make a conversion function between one and the other and use it in lots of places. Let me know if it starts taking too much time.

robertbastian commented on components/locid_transform/src/provider/mod.rs:
```diff
@@ -87,3 +89,24 @@ pub struct StrStrPair<'a>(
+    databake(path = icu_locid_transform::provider),
+)]
+/// A triplet of strings with a EncodeAsVarULE implementation.
+pub struct StrStrStrPair<'a>(
```
Is this type required? `Language` is `Sized` and therefore `ULE`, so `(Language, StrStrPair)` should work?

zbraniecki replied:
I got:
```
179 | #[derive(Debug)]
    |          ----- in this derive macro expansion
...
184 |     pub language_variants: VarZeroVec<'data, UnvalidatedLanguageVariantsPai...
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `VarULE` is not implemented for `(icu_locid::subtags::Language, StrStrPairVarULE)`
```

sffc replied:
I'm fine adding this impl or @robertbastian's suggested impl, but I wanted to observe that this is a misnomer. It is a `LangaugeStrStr`, not a `StrStrStr`.

robertbastian replied:
Agreed on the misnomer

sffc replied:
This misnomer is still not fixed

zbraniecki commented:
> It's a good exercise on how to update our data structs in a forwards-compatible way

I agree. My question is - there are two sides of AliasesV1/V2 - production (datagen) and consumption.

Do I need to datagen AliasesV1 and AliasesV2, or is it sufficient to produce AliasesV2, but ensure I can consume AliasesV1 and AliasesV2?

sffc commented:
> > It's a good exercise on how to update our data structs in a forwards-compatible way
> 
> I agree. My question is - there are two sides of AliasesV1/V2 - production (datagen) and consumption.
> 
> Do I need to datagen AliasesV1 and AliasesV2, or is it sufficient to produce AliasesV2, but ensure I can consume AliasesV1 and AliasesV2?

We're _adding_ a _new_ key; the old key should continue to work. That is to say, Datagen needs to be able to produce AliasesV1 and AliasesV2 so that new datagen can make data for old ICU4X (new data, old code).

https://github.com/unicode-org/icu4x/blob/main/docs/process/data_versioning.md#iii-data-file-versioning

> Old Code, New Data: Datagen can add obsolete resource keys to the data file in order to support a particular minimum version of ICU4X.

sffc commented:
An architecture I would consider is one where you generate AliasesV1 and have a conversion function to AliasesV2, which is used in two places: in datagen and in the compatibility constructors.

You could instead generate AliasesV2, and then you need conversion constructors for both directions.

zbraniecki commented:
@sffc I may need a bit more detailed guidance. I implemented (temporarily) `From<AliasesV1> for AliasesV2` and reverse and added ability to generate AliasesV1 in datagen.

The result is that now datagen generates AliasesV1 alongside AliasesV2, which... hmm, I guess it allows newer data to be loaded into older ICU4X, right?

But now if we want to accomplish the reverse (allow newer ICU4X to read older data), we need to make the constructor able to take AliasesV1 and convert into AliasesV2 for use, right? I'm not sure how to handle this part.

(@robertbastian - I see your comment, I'll experiment with that once I have the conversion working)

sffc commented:
You need to take the with_any_provider and with_buffer_provider out of the macros and write them to read either payload. Try to read V2, and if that fails, read V1 and convert. There is an example of how to do this in LocaleExpander.

sffc commented:
In that file,

- `pub fn try_new_unstable` takes only the modern keys
- `fn try_new_compat` accepts both the legacy and the modern keys
- `try_new_with_any_provider` and `try_new_with_buffer_provider` have manual implementations that call `try_new_compat`, rather than the default macro-generated impls that call `try_new_unstable`
- There are tests at the bottom of the file to make sure the compat function works

robertbastian commented on components/locid_transform/data/config.json:
```diff
@@ -5,6 +5,7 @@
       "fallback/parents@1",
       "fallback/supplement/co@1",
       "locid_transform/aliases@1",
```
compiled data should not carry both structs, remove v1

zbraniecki replied:
If I remove it, I lose ability to test for aliasesv1 vs aliasesv2. Do you have a suggestion how to workaround it?

robertbastian replied:
Yes, compile a blob provider into `tests/data` and use that for tests. I think we do this in this crate already. As this is a backward compatibility test that's not something that will have to be updated again. Datetime does this for example.

zbraniecki replied:
done.

robertbastian commented on components/locid_transform/src/provider/canonicalizer.rs:
```diff
@@ -79,3 +79,143 @@ pub struct AliasesV1<'data> {
     pub subdivision: ZeroMap<'data, UnvalidatedSubdivision, SemivalidatedSubdivision>,
 }
+
+impl<'data> From<AliasesV2<'data>> for AliasesV1<'data> {
```
Can this be `#[cfg(feature = "datagen")]`?

robertbastian commented on components/locid_transform/src/provider/canonicalizer.rs:
```diff
@@ -79,3 +79,143 @@ pub struct AliasesV1<'data> {
+                #[allow(clippy::expect_used)]
+                let (lang, variant) =
+                    v.0.split_once('-')
+                        .expect("Each pair should be language-variant");
```
AliasesV1 could be garbage, you can't expect this.

zbraniecki replied:
should I throw here?

sffc replied:
You should change the `expect` to a `debug_assert` and then return a GIGO value in accordance with the style guide (https://github.com/unicode-org/icu4x/blob/main/docs/design/data_safety.md)

robertbastian replied:
You can either make this a `TryFrom` and throw in the constructor that uses this, or do garbage-in garbage-out

sffc replied:
Yeah, TryFrom might be better here.

robertbastian reviewed (commented):
Mainly looking at data infra, can give the algorithm an in-depth review if you want as well

sffc commented on components/locid_transform/src/canonicalizer.rs:
```diff
@@ -266,24 +260,46 @@ impl LocaleCanonicalizer {
+        Ok(Self { aliases, expander })
     }
 
     icu_provider::gen_any_buffer_data_constructors!(
```
Issue: you'll need to unpack this macro. It's not too hard though since you can use `icu_provider::gen_any_buffer_unstable_docs!`.

zbraniecki replied:
Why do I need to unpack it?

sffc replied:
Code that calls `LocaleCanonicalizer::try_new_with_expander_with_buffer_provider`, they should get the compat constructor, but this macro makes it call the non-compat constructor.

sffc commented on components/locid_transform/src/canonicalizer.rs:
```diff
@@ -230,25 +224,25 @@ impl LocaleCanonicalizer {
+    ) -> Result<Self, LocaleTransformError> {
         let expander = LocaleExpander::try_new_with_buffer_provider(provider)?;
-        Self::try_new_with_expander_unstable(&provider.as_deserializing(), expander)
+        Self::try_new_with_expander_compat(&provider.as_deserializing(), expander)
```
Please change this to call `try_new_with_expander_with_buffer_provider` instead.

zbraniecki replied:
This fails bound check:
```
239 |         Self::try_new_with_expander_with_buffer_provider(&provider.as_deserializing(), expander)
    |         ------------------------------------------------ ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `icu_provider::BufferProvider` is not implemented for `DeserializingBufferProvider<'_, impl BufferProvider + ?Sized>`
    |         |
    |         required by a bound introduced by this call
```

Is it why you asked me to unpack the macro above?

sffc replied:
Take out the `.as_deserializing()` when you call `_with_buffer_provider`

sffc commented on components/locid_transform/src/canonicalizer.rs:
```diff
@@ -230,25 +224,25 @@ impl LocaleCanonicalizer {
+    ) -> Result<Self, LocaleTransformError> {
         let expander = LocaleExpander::try_new_with_any_provider(provider)?;
-        Self::try_new_with_expander_unstable(&provider.as_downcasting(), expander)
+        Self::try_new_with_expander_compat(&provider.as_downcasting(), expander)
```
Please change this to call `try_new_with_expander_with_any_provider` instead.

sffc replied:
^

sffc reviewed (changes requested):
Praise: quite good overall! Just putting down some minor things.

I also didn't look super closely at the canonicalization algorithm part.

zbraniecki commented:
Rebased on top of `main`. Latest perf numbers:

```
uncanonicalized/clone   time:   [1.3410 µs 1.3534 µs 1.3691 µs]
                        change: [+1.3495% +1.8958% +2.5688%] (p = 0.00 < 0.05)
                        Performance has regressed.
uncanonicalized/canonicalize
                        time:   [11.030 µs 11.055 µs 11.085 µs]
                        change: [-45.701% -45.096% -44.633%] (p = 0.00 < 0.05)
                        Performance has improved.

canonicalized/clone     time:   [480.75 ns 481.81 ns 483.24 ns]
                        change: [-0.4505% -0.0310% +0.3906%] (p = 0.89 > 0.05)
                        No change in performance detected.
canonicalized/canonicalize
                        time:   [1.3728 µs 1.3753 µs 1.3785 µs]
                        change: [-6.8665% -6.1423% -5.5612%] (p = 0.00 < 0.05)
                        Performance has improved.

likelysubtags/maximize  time:   [1.0777 µs 1.0794 µs 1.0814 µs]
                        change: [-4.0919% -3.6192% -3.2250%] (p = 0.00 < 0.05)
                        Performance has improved.
```

robertbastian commented:
Please use merge commits once a PR is out for review, it's impossible to get a diff since my last review.

zbraniecki commented:
> Please use merge commits once a PR is out for review, it's impossible to get a diff since my last review.

I generally try to follow that. In this case tho, big parts of the diff were baked data changes and conflicts. Reapplying the PR onto main was simpler and it unfortunately lost the history. I apologize for the churn.

sffc reviewed (commented):
(replied in comment threads)

zbraniecki commented:
@sffc Applied your feedback. Please take a look at the latest stack patch.

sffc commented on components/locid_transform/src/canonicalizer.rs:
```diff
@@ -224,25 +218,25 @@ impl LocaleCanonicalizer {
     pub fn try_new_with_buffer_provider(
-        provider: &(impl BufferProvider + ?Sized),
-    ) -> Result<LocaleCanonicalizer, LocaleTransformError> {
+        provider: &(impl BufferProvider + AnyProvider + ?Sized),
```
```suggestion
        provider: &(impl BufferProvider + ?Sized),
```

sffc commented on components/locid_transform/src/canonicalizer.rs:
```diff
@@ -260,39 +254,74 @@ impl LocaleCanonicalizer {
 
-    icu_provider::gen_any_buffer_data_constructors!(
-        locale: skip,
+    #[cfg(feature = "compiled_data")]
+    #[cfg(skip)]
+    #[doc = ""]
+    #[doc = " ✨ *Enabled with the `compiled_data` Cargo feature.*"]
+    #[doc = ""]
+    #[doc = " [📚 Help choosing a constructor](icu_provider::constructors)"]
```
Nit: Use normal doc strings where you can.

```
        /// Creates a new [`LocaleCanonicalizer`] from compiled data and a [`LocaleExpander`].
        ///
        /// ✨ *Enabled with the `compiled_data` Cargo feature.*
        ///
        /// [📚 Help choosing a constructor](icu_provider::constructors)
```

sffc commented on components/locid_transform/src/canonicalizer.rs:
```diff
@@ -224,25 +218,25 @@ impl LocaleCanonicalizer {
+    ) -> Result<Self, LocaleTransformError> {
         let expander = LocaleExpander::try_new_with_any_provider(provider)?;
-        Self::try_new_with_expander_unstable(&provider.as_downcasting(), expander)
+        Self::try_new_with_expander_with_any_provider(&provider, expander)
```
Thought: This is correct but I guess it's the equivalent of `try_new_with_expander_compat(&provider.as_downcasting())`

sffc reviewed (commented):
Good with one little typo

sffc commented on components/locid_transform/src/canonicalizer.rs:
```diff
@@ -260,39 +254,74 @@ impl LocaleCanonicalizer {
+        Self::try_new_with_expander_unstable(&crate::provider::Baked, options)
+    }
+
+    #[doc = icu_provider::gen_any_buffer_unstable_docs!(ANY, Self::try_new_with_expander_with_buffer_provider)]
+    pub fn try_new_with_expander_with_any_provider(
```
Nit: I think the arguments to this macro might be wrong. Please double-check. Also below.

robertbastian commented on components/locid_transform/src/provider/mod.rs:
```diff
@@ -59,6 +60,7 @@ use icu_provider::prelude::*;
 /// The latest minimum set of keys required by this component.
 pub const KEYS: &[DataKey] = &[
     AliasesV1Marker::KEY,
+    AliasesV2Marker::KEY,
```
Remove the V1 key? We don't want both in baked data.

robertbastian commented on components/locid_transform/src/canonicalizer.rs:
```diff
@@ -260,39 +254,74 @@ impl LocaleCanonicalizer {
-    icu_provider::gen_any_buffer_data_constructors!(
-        locale: skip,
+    #[cfg(feature = "compiled_data")]
+    #[cfg(skip)]
```
we're using this in macros to disable the method, but you probably don't want that

zbraniecki replied:
Yeah, I don't need this method here. It's explicitly listed above. Removed.

robertbastian commented on components/locid_transform/src/canonicalizer.rs:
```diff
@@ -260,39 +254,74 @@ impl LocaleCanonicalizer {
+    #[doc = " ✨ *Enabled with the `compiled_data` Cargo feature.*"]
+    #[doc = ""]
+    #[doc = " [📚 Help choosing a constructor](icu_provider::constructors)"]
+    pub fn new_with_expander(options: LocaleExpander) -> Result<Self, LocaleTransformError> {
```
This should be infallible. Use the singleton instead of going through `_unstable`

robertbastian replied:
argument name should be `expander`
